### PR TITLE
[KYUUBI #5317] [Bug] Hive Connector throws NotSerializableException on reading Hive Avro partitioned table

### PR DESCRIPTION
**Pull request by zhaomin1423**
### _Why are the changes needed?_

close https://github.com/apache/kyuubi/issues/5317#issue-1904751001

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_

No

### CODE REVIEW

File: extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala [hunk 1]
```diff
@@ -33,12 +33,15 @@ class HiveQuerySuite extends KyuubiHiveTest {
...
+  def withTempPartitionedTable(
+      spark: SparkSession,
+      table: String,
+      format: String = "PARQUET")(f: => Unit): Unit = {
     spark.sql(
       s"""
          | CREATE TABLE IF NOT EXISTS
          | $table (id String, year String, month string)
-         | USING PARQUET
+         | USING $format
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala:44]: I'm not sure how "USING" would be handled in this connector, I verified with `STORED AS` instead

File: extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala [hunk 2]
```diff
@@ -186,4 +189,19 @@ class HiveQuerySuite extends KyuubiHiveTest {
       }
     }
   }
+
+  test("read partitioned avro table") {
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala:193]: let's also add test for un-partitioned serde table

Comment (zhaomin1423) [reply extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala:208]: Done, and add other formats.

File: extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala (hunk 1 above)
Comment (zhaomin1423) [reply extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala]: Both work and also add unit tests

**Comment by Yikf** [comment]
@zhaomin1423 Thanks, LGTM

File: extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala (hunk 1 above)
Comment (Yikf) [reply extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala]: As far as i known, This point depend on the implementation of the catalog. v2HiveCatalog creates tables and queries them, in fact, STORED AS & USING are the same thing.

Comment (zhaomin1423) [reply extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala]: There is a small difference, the format of text does not work when STORED AS is used, only textfile, but USING work well.

**Comment by ulysses-you** [comment]
lgtm

File: extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala [hunk 3]
```diff
@@ -22,23 +22,41 @@ import org.apache.spark.sql.catalyst.analysis.UnresolvedRelation
...
+  def withTempPartitionedTable(
+      spark: SparkSession,
+      table: String,
+      format: String = "PARQUET",
+      hiveTable: Boolean = false)(f: => Unit): Unit = {
+    val clause = if (hiveTable) {
+      "STORED AS"
+    } else {
+      "USING"
+    }
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala:54]: let's inline it

File: extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala (hunk 1 above)
Comment (Yikf) [reply extensions/spark/kyuubi-spark-connector-hive/src/test/scala/org/apache/kyuubi/spark/connector/hive/HiveQuerySuite.scala]: this is caused by the parsing of spark, In spark, text file format is `text`, hive stored as is `textfile`

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/read/HivePartitionReaderFactory.scala [hunk 4]
```diff
@@ -64,17 +65,22 @@ case class HivePartitionReaderFactory(
     val filePartition = partition.asInstanceOf[FilePartition]
     val iter: Iterator[HivePartitionedFileReader[InternalRow]] =
       filePartition.files.toIterator.map { file =>
-        val bindHivePart = partFileToHivePart.getOrElse(file, null)
+        val bindHivePart = partFileToHivePart.get(file)
+        val hivePartition = if (bindHivePart.isDefined) {
+          HiveClientImpl.toHivePartition(bindHivePart.get, hiveTable)
+        } else {
+          null
+        }
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/read/HivePartitionReaderFactory.scala:73]:
```suggestion
        bindHivePart.map(HiveClientImpl.toHivePartition(_, hiveTable)).orNull
```

**Review by pan3793 — APPROVED** [review]
LGTM, except to small nits

**Comment by zhaomin1423** [comment]
> LGTM, except to small nits

Update, thanks.

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/read/HivePartitionReaderFactory.scala [hunk 5]
```diff
@@ -64,17 +65,18 @@ case class HivePartitionReaderFactory(
...
           new PartitionReaderWithPartitionValues(
             HivePartitionedReader(
               file,
-              buildReaderInternal(file, bindHivePart),
+              buildReaderInternal(file, hivePartition),
               tableDesc,
               broadcastHiveConf,
               nonPartitionReadDataKeys,
-              bindHivePart,
+              hivePartition,
```
[the remark this answers is not in the document]
Comment (zhaomin1423) [reply extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/read/HivePartitionReaderFactory.scala:79]: Done

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/read/HivePartitionReaderFactory.scala [hunk 6]
```diff
@@ -99,11 +99,11 @@ case class HivePartitionReaderFactory(
 
   private def createPartitionWritableReader[T](
       file: PartitionedFile,
-      bindPartition: HivePartition): Iterator[Writable] = {
+      bindPartition: Option[HivePartition]): Iterator[Writable] = {
     // Obtain binding HivePartition from input partitioned file
     val partDesc =
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/read/HivePartitionReaderFactory.scala:104]: use `Option` too

**Review by cxzl25 — APPROVED** [review]

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/read/HivePartitionReaderFactory.scala (hunk 6 above)
Comment (zhaomin1423) [reply extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/read/HivePartitionReaderFactory.scala]: done

**Comment by pan3793** [comment]
The last two commits only change the code style, merging to master/1.8

**Pull request closed without merging**